File: 0502-ipo/0502-ipo.java
class Pair{
    int cap;
    int prof;
    Pair(int cap, int prof){
        this.cap = cap;
        this.prof = prof;
    }
}
class Solution {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        int n = profits.length;
        Pair[] p = new Pair[n];
        for(int i = 0; i < n; i++){
            p[i] = new Pair(capital[i], profits[i]);
        }

        Arrays.sort(p, (a,b)-> a.cap - b.cap);


        PriorityQueue<Pair> pq = new PriorityQueue<>((a,b)->Integer.compare(b.prof, a.prof));

        int idx = 0;
        for(int i = 0; i < k;i++){
            while(idx < n && w >= p[idx].cap){
                pq.add(p[idx]);
                idx++;
            }

            if(pq.isEmpty()){
                return w;
            }
            w += pq.poll().prof;

        }
        return w;
    }
}

File: LeetCode/Hard/0502-ipo/0502-ipo.cpp
class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        int n = profits.size();
        vector<pair<int, int>> projects;

        // {(c, p), (c, p)....}
        for(int i = 0; i < n; i++){
            projects.push_back({capital[i], profits[i]});
        }
        sort(projects.begin(), projects.end()); // sorted according to capital

        priority_queue<int> pq;
        int idx = 0;
        while(k--){
            while(idx < n && projects[idx].first <= w) pq.push(projects[idx++].second);
            // check all the projects which u can do now and push the profits into max heap
            if(pq.empty()) break;
            // try to do project which gives max profit

            w += pq.top();
            pq.pop();
        }
        return w;


    }
};